### ISSUE TITLE
Workflow по Pull Request в ядро

### ISSUE THREAD

**Issue by b2z** [RUJL/rujl.github.io#7]
Обсудили с @Septdir. Вырисовывается такой вариант:
- делается PR в наш форк репозиторий (fork) Joomla! CMS
- как только все вопросы улажены (оформлен текст PR, протестировано), то мерджим в отдельную ветку
- далее эту ветку проталкиваем как PR в основной репозиторий (upstream) Joomla! CMS, но уже с английским описанием
https://help.github.com/articles/creating-a-pull-request-from-a-fork/
- если необходимо внести изменения, то они вносятся в ветку нашего форк репозитория Joomla! CMS

**Comment by Septdir** [comment]
Начинаю тестировать!

## Создание форка
Если у вас **нет** форка `joomla/joomla-cms` то можно сделать форк  ` JoomlaRu/joomla-cms` 
Если у вас **уже есть** форк `joomla/joomla-cms`, то еще одни форк делать не нужно, да и не получиться.

**Comment by Septdir** [comment]
## Код PR
Все тоже самое что в Joomla можно прям с офф доки скопировать, но добавить строку  
**Не забудьте сделать новую ветку перед началом работы над PR**

## Создание самого PR
Схема так же. Самое важное выбрать правильную ветку для сравнения **JoomlaRu/joomla-cms**
Если форк делался с `JoomlaRu/joomla-cms` то она будет по умолчанию.  
_Иначе можно как я вчера отправить PR в Joomla_

![pr-1](https://user-images.githubusercontent.com/18382023/47656814-ee0ed580-dba0-11e8-8acc-146d8d0b6ce8.png)
![pr-2](https://user-images.githubusercontent.com/18382023/47656821-f1a25c80-dba0-11e8-89e0-55070fb512c2.png)

**Comment by Septdir** [comment]
## Блокируем ветки!
**Самое главное заблокировать ветки**
Если ветка не заблокирована то PR можно будет мерждить в ветку `staging` А нам этого не нужно.
![pr-m-1](https://user-images.githubusercontent.com/18382023/47657865-f831d380-dba2-11e8-8dfd-796593e6af83.png)

Поэтому блокируем ветку.
Уже давно искал как просто полностью заблокировать merge в ветку, но не нашел. Поэтому вот такой вариант.
![pr-m-2](https://user-images.githubusercontent.com/18382023/47658510-2c59c400-dba4-11e8-9268-0900d7f9b092.png)
И получаем вот такую блокировку
![pr-m-3](https://user-images.githubusercontent.com/18382023/47658594-4dbab000-dba4-11e8-80ac-d5e0e533936f.png)

**Comment by Septdir** [comment]
## Создаем отдельную ветку для PR
Создаем новую ветку. С `staging` все просто. А вот если надо с другой ветки то тут либо через локальную репу + команду
```
git checkout -b {NEW_branch_name} {SOURCE_branch_name}
git push origin {NEW_branch_name}
```
Либо через Github Desktop  

Либо поменять в настройках ветку по умолчанию и тогда можно будет сделать прямо на github.

Именовать предлагаю по номеру pr к примеру `path-23`, где 23 id PR

**Comment by Septdir** [comment]
## Меняем ветку
После создания ветки меняем ветку в PR
![pr-m-4](https://user-images.githubusercontent.com/18382023/47659813-ccb0e800-dba6-11e8-846b-6969cf418216.png)

И и тогда PR можно мерджить.

![pr-m-5](https://user-images.githubusercontent.com/18382023/47659818-cfabd880-dba6-11e8-8f99-e6f6fad1cd99.png)

**Comment by Septdir** [comment]
Остаются тесты
`Travis CI`  и `appveyor` то не проблема, я конечно не спец, но в крайнем случае настроить смогу. А вот drone лично я ни разу не настаивал. И Human тест нужен.  

Может с кем нибудь из Joomla посоветоваться?  
_Цель у нас благая, глядишь подсобят, если не делом то хоть советом и гайдами =)

**Comment by b2z** [comment]
Так, немного сумбурно, но в целом идея понятна.  Нормально доку сделаем и будет гуд. 

Я только вот что не понял - если необходимо внести изменения, то они вносятся в ветку нашего форк репозитория Joomla! CMS. Это точно будет работать?

**Comment by Septdir** [comment]
### Насчет сумбурно. 
Согласен полностью. Пока это скорее не более чем заметки, причем вторая часть вообще касается только тех кто будет модерировать.

Так, что наверное надо будет 2 руководства сделать.

### Насчёт изменений.  
Тут все работает как со стандартными PR.

Если человек хочет что-то изменить, до того как мы его смерджири к себе в отдельную ветку, то делает это как обычно у себя в форке. 

Если же он или кто либо еще хочет сделать изменении после того мы создали PR в оригинальный репозиторий то это делается через PR в созданную нами ветку.

Вот пример PR с PR внутри.
https://github.com/joomla/joomla-cms/pull/20229

К тому же если исправления мелкие, ну типа пропущенного пробела или просто доработки, то мы можем и сами это сделать  в ново созданной ветки которая связана с PR. Заодно и человеку подсобим,

---
Завтра попробую сделать чтобы сообщения об успешном тестировании можно было прямо из патч тестера отправлять.

Если получиться можно будет PR сделать чтобы эту функцию добавили, если сочтут нужной конечно. А пока плагин сделать. 

___

P.S Если дело пойдет, то можно будет весь этот процесс с созданием веток и PR в оригинал автоматизировать.  Вплоть до такого.  

В админе нажал на одну копку: 
Create Branch (сохраняем название исконной ветки в переменной) > Merge PR (в новую ветку) > Create PR(в Joomla/joomla-cms можно через трайслейт прогнать, потом поправить)  
*Насчет создания PR в Joomla не уверен но судя по доке можно**
